### PR TITLE
FIX:Matlab path search on macOS

### PR DESCRIPTION
**Pull request by matl-hsk**
**Description**
- fixes creation of the path were matlab is installed on macOS
- fixes search for latest matlab version

**Related Issues**
This pull-request fixes issue #6387

**Tasks**
Add the list of tasks of this PR.
  - [x] Test in macOS
  - [x] Test in Windows
  - [ ] Test in Linux

**Documentation**
No change in documentation.

**Additional context**
See issue #6387

### CODE REVIEW

File: src/controller/launcher/webots_controller.c [hunk 1]
```diff
@@ -169,54 +169,61 @@ static bool get_webots_home() {
...
   }
-#endif
+
   // Get latest available Matlab version
   char *latest_version = NULL;
   while ((directory_entry = readdir(directory)) != NULL) {
     const size_t directory_name_size = strlen(directory_entry->d_name) + 1;
     if (strncmp(matlab_version_wc, directory_entry->d_name, strlen(matlab_version_wc)) == 0) {
       if (!latest_version)
+      {
```
Comment (omichel) [inline src/controller/launcher/webots_controller.c:206]:
```suggestion
      if (!latest_version) {
```

**Review by omichel — CHANGES_REQUESTED** [review]
There are some clang-format issues.

File: src/controller/launcher/webots_controller.c [hunk 2]
```diff
@@ -169,54 +169,61 @@ static bool get_webots_home() {
...
         latest_version = malloc(directory_name_size);
+        strncpy(latest_version, directory_entry->d_name, directory_name_size);
+      }
       else if (strcmp(latest_version, directory_entry->d_name) < 0)
+      {
```
Comment (omichel) [inline src/controller/launcher/webots_controller.c:211]:
```suggestion
      else if (strcmp(latest_version, directory_entry->d_name) < 0) {
```

File: src/controller/launcher/webots_controller.c [hunk 3]
```diff
@@ -169,54 +169,61 @@ static bool get_webots_home() {
...
-#ifdef __APPLE__
-  const size_t matlab_path_size = snprintf(NULL, 0, "%s%s", matlab_directory, latest_version) + 1;
-  matlab_path = malloc(matlab_path_size);
-  sprintf(matlab_path, "%s%s", matlab_directory, latest_version);
-#else
   const size_t matlab_path_size = snprintf(NULL, 0, "%s%s%s", matlab_directory, latest_version, matlab_exec_suffix) + 1;
   matlab_path = malloc(matlab_path_size);
   sprintf(matlab_path, "%s%s%s", matlab_directory, latest_version, matlab_exec_suffix);
-#endif
+  printf("Using matlab path %s\n",matlab_path);
```
Comment (omichel) [inline src/controller/launcher/webots_controller.c:226]: Is that a debug statement, can we remove it?
```suggestion
```

Comment (matl-hsk) [reply src/controller/launcher/webots_controller.c:226]: I left it intentionally, but wen can remove it, as well.

**Comment by nilsjor** [comment]
Whoops I came on and made your review "outdated". Sorry about that...
I though it would be a suggestion like yours, but it became a commit. Anyways, I fixed the clang-format. 

I can confirm it works on Windows! Only thing to address is the debug info about the found path (https://github.com/cyberbotics/webots/pull/6390#discussion_r1335766194)

File: src/controller/launcher/webots_controller.c [hunk 4]
```diff
@@ -223,7 +220,7 @@ static bool get_matlab_path() {
   const size_t matlab_path_size = snprintf(NULL, 0, "%s%s%s", matlab_directory, latest_version, matlab_exec_suffix) + 1;
   matlab_path = malloc(matlab_path_size);
   sprintf(matlab_path, "%s%s%s", matlab_directory, latest_version, matlab_exec_suffix);
-  printf("Using matlab path %s\n",matlab_path);
+  printf("MATLAB instance found at: %s\n", matlab_path);
```
Comment (matl-hsk) [inline src/controller/launcher/webots_controller.c:223]: This

Comment (nilsjor) [reply src/controller/launcher/webots_controller.c:223]: This is a good point. I agree! However, 
- MATLAB should be capitalized in order to be consistent. 
- The colon should be left in order to be consistent with
`Using prototype file: C:\msys64\tmp\protofile_matlab_2021a_webots_R2023b_revision.m`
- The space between args (col 36) will cause an error in clang-format

Comment (nilsjor) [reply src/controller/launcher/webots_controller.c:223]: `printf("Using the latest available MATLAB instance: %s\n", matlab_path);`

File: src/controller/launcher/webots_controller.c [hunk 5]
```diff
@@ -169,54 +169,58 @@ static bool get_webots_home() {
...
-#ifdef __APPLE__
-  const size_t matlab_path_size = snprintf(NULL, 0, "%s%s", matlab_directory, latest_version) + 1;
-  matlab_path = malloc(matlab_path_size);
-  sprintf(matlab_path, "%s%s", matlab_directory, latest_version);
-#else
   const size_t matlab_path_size = snprintf(NULL, 0, "%s%s%s", matlab_directory, latest_version, matlab_exec_suffix) + 1;
   matlab_path = malloc(matlab_path_size);
   sprintf(matlab_path, "%s%s%s", matlab_directory, latest_version, matlab_exec_suffix);
-#endif
+  printf("MATLAB instance found at: %s\n", matlab_path);
```
Comment (matl-hsk) [inline src/controller/launcher/webots_controller.c:223]:
```suggestion
  printf("Using MATLAB instance: %s\n", matlab_path);
```

**Review by nilsjor — APPROVED** [review]

File: src/controller/launcher/webots_controller.c [hunk 6]
```diff
@@ -169,54 +169,58 @@ static bool get_webots_home() {
...
   while ((directory_entry = readdir(directory)) != NULL) {
     const size_t directory_name_size = strlen(directory_entry->d_name) + 1;
     if (strncmp(matlab_version_wc, directory_entry->d_name, strlen(matlab_version_wc)) == 0) {
-      if (!latest_version)
+      if (!latest_version) {
         latest_version = malloc(directory_name_size);
-      else if (strcmp(latest_version, directory_entry->d_name) < 0)
+        strncpy(latest_version, directory_entry->d_name, directory_name_size);
+      } else if (strcmp(latest_version, directory_entry->d_name) < 0) {
         memset(latest_version, '\0', directory_name_size);
```
Comment (omichel) [inline src/controller/launcher/webots_controller.c:209]: I believe there is a potential bug here: if `directory_name_size` becomes bigger on a subsequent iteration (which is possible if we have some file or sub-folder with a long name), we may perform a buffer overflow by copying `\0` beyond the boundary of the allocated memory segment, causing potentially a crash. Thus, I believe the `latest_version` buffer should be reallocated to fit the exact size needed for the copy.
```suggestion
        latest_version = realloc(latest_version, directory_name_size);
```

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/controller/launcher/webots_controller.c (hunk 6 above)
Comment (matl-hsk) [reply src/controller/launcher/webots_controller.c:209]: agree

**Review by omichel — APPROVED** [review]

**Comment by matl-hsk** [comment]
Do you have someone to test it on linux or shall I go ahead and merge?

**Pull request merged**